### PR TITLE
Added class TGraphMultiErrors for graph with multidimensional y-Errors

### PR DESCRIPTION
**Pull request by Triple-S**
Hello,

First of all sorry for all the mess I made with these pull requests. This is my first try and I did not really understood how GitHub works therefore my previous pull request was this messy and I decided to withdraw it and clean everything up before I request it again. And here we are, I ran some additional bugfixing and the ROOT tests which succeeded. Furthermore I tested my new class and everything I tried worked so I think now it's time to request the pull again.

**Explanation**
My implementation is named TGraphMultiErrors and is basically a TGraphAsymmErrors with the possibility to include an unlimited amount of different types of y-Errors.
The whole functionality and naming convention of my class is strongly based on the TGraphAsymmErrors class. All functionalities of the TGraphAsymmErrors have been reimplemented. Furthermore like any other TGraph class it inherits from TGraph and has therefore all of it’s functionalities.

I attached an example plot as a png picture. It's macro to be produced is included in the description part of the TGraphPainter class.

**Differences to TGraphAsymmErrors**
Everywhere TGraphAsymmErrors uses arrays for the y-Errors, my class uses two dimensional arrays (Double_t**) of which the first dimension represents the index of the error dimension and the second one the point. It is assumed that the first y-Error is the statistical one and all following are systematical ones. Therefore when the statistical errors are being computed by a function everything applies to only the first error dimension.
In case a single error on y is required (Like for fitting) the errors are being summed according to the value of fSumErrorsMode which can be at the moment TGraphMultiErrors::kOnlyFirst for only returning the error of the first error type (usally the statistic one), TGraphMultiErrors::kSquareSum for returning the squared sum of the errors of all types (Gaussian propagation) or TGraphMultiErrors::kSum for just returning the sum of the errors from all error types.

**Drawing / Painting**
The painting is being performed by the TGraphPainter class like for any other classes. Therefore I added the function TGraphPainter::PaintGraphMultiErrors and included a corresponding line in the TGraphPainter::PaintHelper function.

**Drawing / Painting options**
To be able to define different drawing options for the multiple error dimensions the option string can consist of multiple blocks separated by semicolons. The painting method separates the options in multiple options for each y error dimension and one for basic painting and the x errors. In case there are <= NErrorDimensions blocks given, the first block is used for basic painting and for the first dimension of y errors. Y error dimensions without an individual block are painted with “”.
Tn case there are NErrorDimensions + 1 given, the first block is used only for basic painting. The remaining blocks are distributed over the y error dimensions In case there are > NErrorDimensions + 1 blocks given an error is returned. All drawing options of TGraphAsymmErrors are implemented. For explanation of the drawing options see TGraphPainter::PaintGraphAsymmErrors.
Additionally one can specify x0 or y0 to skip points with x = 0 or y = 0 The Draw option s=%f can be used to multiply the corresponding x errors with %f. This is useful in combination with e.g. option 2 if the boxes are supposed to have only half the width of the bin, in this case the option would be s=0.5. Per default the Fill and Line Styles of the Graph are being used for all error dimensions. To use the specific ones add the draw option s to the first block.
In my example plot the drawing options are “a p s ; ; 5 s=0.5” which means the graph is drawn with Axes, Markers and individual Fill and Line attributes for all error dimensions for the general part. The first error dimension (statistical errors) is drawn with default settings. The second error dimension (systematic errors) is drawn with boxes of half the width of the X-Errors.

**Individual Fill and Line Attributes for each error dimension**
This is being realized by overloading the Getter and Setter functions of TAttFill and TAttLine with an additional dimension parameter. Furthermore there are Getter and Setter functions for full TAttFill and TAttLine objects to get / set all attributes at once.

To keep full compatibility with TGraphAsymmErrors by default the TAttFill and TAttLine attributes of the object are being used for all error dimensions. To activate using the individual ones add the Draw option “S” to the first block of the Draw options like in my example.

All the best,
Simon Spies

I hope you find my class as useful as I do and vote for it to become part of ROOT in the future. :smile:
example
example

### CODE REVIEW

**Comment by Triple-S** [comment]
@phsft-bot build

**Comment by Triple-S** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by Triple-S** [comment]
Hmm, strange that the build fails, but the error seems to be not connected to the PR:

> 22:53:26 < HTTP/1.1 401 Authorization Required
> 22:53:26 < Server: GitHub Babel 2.0
> 22:53:26 < Content-Type: text/plain
> 22:53:26 < Content-Length: 21
> 22:53:26 < WWW-Authenticate: Basic realm="GitHub"
> 22:53:26 < X-Frame-Options: DENY
> 22:53:26 < X-GitHub-Request-Id: F8B0:30DD2:1EF7253:2E3CE06:5C79A9D6
> 22:53:26 < 
> 22:53:26 * Connection #0 to host github.com left intact
> 02:52:59 Cancelling nested steps due to timeout

A few seconds before a similar request succeded. I found this line a few lines before the fail:

> 22:53:26 > GET /Triple-S/roottest.git/info/refs?service=git-upload-pack HTTP/1.1

At this moment I had no fork of roottest, so this might be the reason the request failed. I just added a fork of roottest so hopefully this error will not occur again.

**Comment by couet** [comment]
@phsft-bot build

**Comment by couet** [comment]
@lmoneta Can you comment on this one ?

File: hist/hist/inc/TGraphMultiErrors.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+   Double_t *GetEXhigh() const { return fExH; }
+   Double_t *GetEYlow() const { return fEyLSum; }
+   Double_t *GetEYhigh() const { return fEyHSum; }
+   Double_t *GetEYlow(Int_t dim);
+   Double_t *GetEYhigh(Int_t dim);
+
+   TAttFill *GetAttFill(Int_t dim) const;
+   TAttLine *GetAttLine(Int_t dim) const;
+
+   using TAttFill::GetFillColor;
```
Comment (anerokhi) [inline hist/hist/inc/TGraphMultiErrors.h]: Why did you add this (and other) `using`?

**Comment by Triple-S** [comment]
Thanks for you comments @lmoneta.
I'm sorry for my late reply but since nothing happened here for over a month, I stopped watching.

The reason I used double pointers for the 2D arrays was to have the most low level datatype possible, have full felxibility and stay consistent with the other TGraph classes which are all using pointers for 1D arrays. Nevertheless I understand your concerns since double pointers for arrays are a rather dirty way of programming.

I think it might be a good compromise to use the ROOT array wrapper class TArrayD. Then the two dimensional Arrays could be represented by an 1D array of TArrayD objects.

The functions SetPointEY, SetPointEYL and SetPointEYH were implemented by me to keep full consistency with the TGraphAsymmErrors class. They are overloaded with one version taking the point and an array in which case the array cointains all errors of the named point for the different dimensions. My idea was to alter the TGraphAsymmErrors notation as little as possible so TGraphAsymmErrors::SetPointEY(i, el, eh) would translate to TGraphMultiErrors::SetPointEY(i, &el, &ey) in the case of only one error dimension which would be equal to a TGraphAsymmErrors.

The idea of the function SetDimensionEY, SetDimensionEYL and SetDimensionEYH is basically the same. They set the errors of all points in the given dimension. My idea was to think of a way to convert from multiple TGraphAsymmErrors objects to one single TGraphMultiErrors object like in this small example:
```
TGraphAsymmErrors* statisticErrors = ...
TGraphAsymmErrors* systematicErrors = ...

TGraphMultiErrors* theGraph = new TGraphMultiErrors(statisticErrors->GetN(), 2, statisticErrors->GetX(), statisticErrors->GetY(), statisticErrors->GetEXlow(), statisticErrors->GetEXhigh());
theGraph->SetDimensionEY(0, statisticErrors->GetEYlow(), statisticErrors->GetEYhigh());
theGraph->SetDimensionEY(1, systematicErrors->GetEYlow(), systematicErrors->GetEYhigh());
```

But again I see your point since there is no check whether the passed array has a sufficient size. Maybe here it would also be a good solution to use the wrapper class TArrayD instead since it provides the neccessary information and is rather lightweight.

**Comment by pamputt** [comment]
@couet @lmoneta is there any news about  this pull request? It would be very nice to have this feature in the next ROOT version.

**Comment by couet** [comment]
@lmoneta . It looks fine for me. Is it for you ?

**Comment by Triple-S** [comment]
I'm still waiting for a reply from @lmoneta to my last comment since he had some objections concerning my use of double pointers for two dimensional arrays and passing an array to a function without passing the corresponding size (which can result in a segmentation violation if the user is not creful)

Furthermore, I explained the functions SetPointEY and SetDimensionEY.

I proposed a solution, but before I implement it, I want @lmoneta to check and confirm it, because otherwise I might waste my time.

**Comment by lmoneta** [comment]
Sorry for my late reply. Thank you very much for your answer. 

I think using TArayD is an improvement than using double pointers, but I don't see any reason why not using a more standard std::vector.  
I understand also for the interface your preference for an API similar to the current TGraphAsymmError.

Thank you for explaining the meaning of SetDimensionEY, I think the name is not super clear, I would maybe just call it SetEY. 
One comment on the passed array, I think the pointers should be passed as cont , since they will not be modified. 

One other comment I have , I will not have a method to Set the error dimension. I think it will be better instead a AddError method that will increment by one the dimension and add all the point errors. This I think it will be more useful

**Comment by Triple-S** [comment]
> I think using TArayD is an improvement than using double pointers, but I don't see any reason why not using a more standard std::vector.
> I understand also for the interface your preference for an API similar to the current TGraphAsymmError.

The reason I really don't want to use std::vector here is the streaming of the class. I know that there are ways to stream STL containers, but I don't know how they really work and I refuse to make the functionality of my code rely on something I don't understand.
Furthermore, using ROOT objects provides the advantage that the streaming of the storage class is fully handled by the class itself and therefore there is no need to think about the correct way to stream it. I'm also not 100% sure whether STL containers can be streamed by an automatically built streamer but ROOT classes can be streamed by them for sure.
Last but not least I hold the opinion that ROOT classes should use other ROOT classes anywhere possible, cause we, the programmers, can always change ROOT classes if we need to, but we are not able to change C++ classes.

> Thank you for explaining the meaning of SetDimensionEY, I think the name is not super clear, I would maybe just call it SetEY.
> One comment on the passed array, I think the pointers should be passed as cont , since they will not be modified.

I fully agree on these points.

> One other comment I have , I will not have a method to Set the error dimension. I think it will be better instead a AddError method that will increment by one the dimension and add all the point errors. This I think it will be more useful

This point I don't fully understand. To me it sounds like you propose to add a function like this:
```
void TGraphMultiErrors::AddError(const Double_t *eyL = 0, const Double_t *eyH = 0)
{
   SetNErrorDimensions(fNErrorDimensions + 1);
   if (eyL && eyH)
      SetDimensionEY(fNErrorDimensions - 1, eyL, eyH);
}
```
This would definitly be an useful addition, but I would keep the SetNErrorDimensions function, because otherwise there would be no way to reduce the amount of error dimensions which could be problematic when you want to store the object in a file but not all error dimensions are required. Reducing the amount of dimensions in that case also reduces the amount of storage required, which will be especially important using several graphs or graphs with a lot of points.

**Comment by lmoneta** [comment]
Hi, 

 Thank you for your fast reply. Streaming stl container in ROOT of standard types is easly done as doing for native ROOT objects.
But I understand you have this preference, and since you are the class author we accept this.  

For SetNErrorDimensions  I understand the use case to reduce the dimensionality before storing, but then it would not be better to have the capability to delete a full error dimension with a dedicated function ? 
And correct, I was proposing before having an AddError function 


I think before merging this PR it would be nice to have also two additional things: 

 - one tutorial showing the usage of the class
 - a test program testing the basic functionality. This could be written for example in gtest and included in the hist/hist/test directory where we have other tests of new histogram classes 


Lorenzo

**Comment by Triple-S** [comment]
> Thank you for your fast reply. Streaming stl container in ROOT of standard types is easly done as doing for native ROOT objects.
> But I understand you have this preference, and since you are the class author we accept this.

I will try to figure out how the streaming of STL containers is done, then I don't feel bad about using them, since they also have some advantages.

> For SetNErrorDimensions I understand the use case to reduce the dimensionality before storing, but then it would not be better to have the capability to delete a full error dimension with a dedicated function ?

I also think that might be an even better solution. I will look into it.

> I think before merging this PR it would be nice to have also two additional things:
> 
>     * one tutorial showing the usage of the class
> 
>     * a test program testing the basic functionality. This could be written for example in gtest and included in the hist/hist/test directory where we have other tests of new histogram classes

I will also look into this.

I will probaly need some time to implement and test all the requested changes, especially since I have also some work to do for my PhD, but I see no severe problems at the moment.

Thanks and all the best,
Simon

**Comment by Triple-S** [comment]
OK, please ignore this commit. I just wanted to update my fork to the current state of the root project and did not knew github would create a commit for this.

**Comment by Triple-S** [comment]
Well, actually it was less work than I thought.
I implemented now all the additions, changes and suggestions from @lmoneta

I used now std::vector everywhere it's features are useful. For the rest I stayed with using arrays to stay similar to all other TGraph objects.

I moved the functions GetPointX, GetPointY, SetPointX, SetPointY and GetObjectInfo to the TGraph class because they apply to any TGraph object and only were in my custom class cause I could not implement them in TGraph without using a custom Root installation.

I replaced the funtion SetNErrorDimensions by the functions AddYError and DeleteYError as proposed by @lmoneta. Since I think the meaning of "dimension" might not be 100% clear, I exchanged it to a term claryfing that it corresponds to different y errors.

Besides that I implemented some further features and fixed a memory leak in the function PaintGraphMultiErrors in the TGraphPainter class.

I also added a test macro testing the most critical features like streaming the class, adding / deleting points and y errors and the summation of y errors for fitting with different methods.
Furthermore, I added a small tutorial macro that shows how to use the basic functionalities of the class.

**Comment by Triple-S** [comment]
OK, it seems that Travis CI sometimes request changes that break the functionality. Therefore, I will not continue to implement any changes requested by it. I anyhow think the most important problems are fixed now.

**Comment by Triple-S** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
Hi, 
 Thank you for applying the changes in the code and to provide a tutorial and a test program. 
I have one more comment:
 I still don't like having to pass a double** in the interface (in this case in the constructor). As you see that makes the code not nice (you see that also Travis complained). Since there is no easy solution for this, I would use std::vector<std::vector>> which gives the advantage that you could make it simply directly from a 2D initializer list. Otherwise my second preference is to use a simple double or float pointer and one assumes a proper order of the data. (e.g. raw major). 

I would prefer having std::vector<std::vector<double>> for the contained type instead of std::vector<TArrayD>.   

Thank you again for your changes and I hope we can merge this PR soon. Sorry for my delay in reviewing it 

 Lorenzo
 Thank you if you can apply the ch

**Comment by Triple-S** [comment]
Hi,
I think I have found a solution that fulfilles all requirements.
First of all the good news: I got rid of all double pointers and two dimensional arrays.

I overhauled all the constructors and since I was already at it, I also added constructors which take a name and a title, as it is implemented for histograms. The notations without name and title still persist so the class stays compatible with all other TGraph objects.

For the constructors which take y errors, I implemented one which takes std::vector<std::vector<Double_t>> as prefered by you and one which takes std::vector<TArrayD> which ist the implementation I prefer, since I still think one should use STL containers only when there is no way to achieve the same result with ROOT objects.

Another change I implemented is changing the arrays that store the summed y-errors to mutable because that allows me to implement the functions GetEYlow() and GetEYhigh() inherited from TGraph with the exact same signature which eliminated possible problems with graph objects declared as constant.

I hope I could fulfill all your expectations with these changes.
All the best,
Simon

**Comment by lmoneta** [comment]
Hi

 Thank you Simon for the update and this very nice and useful contribution! 
The changes look fine to me. For my opinion this PR is ready to be merged!

Best, 

Lorenzo

**Comment by Triple-S** [comment]
Ups, sorry, these were my fat fingers on the tablet screen, please ignore the review request.

**Comment by lmoneta** [comment]
The PR is now merged! Thank you very much Simon for your contribution. If we will have some issues in the next days before the release, we will let you know
Cheers
Lorenzo

**Comment by Triple-S** [comment]
Thank you very much!
I'm glad I can finally consider this project closed :D
In case any problems occur, I will help fixing them.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by Triple-S** [comment]
@vgvassilev I'm sorry my class crashed root.
I think I know where the problem comes from and created a new pull request to fix them.
https://github.com/root-project/root/pull/4698
